Source/WebCore/dom/ElementTextDirection.cpp

```diff
@@ -0,0 +1,326 @@
+    return std::nullopt;
+}
+
+std::optional<TextDirection> computeElementAutoTextDirection(const Element& element)
+{
+    if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(element))
+        return computeTextFormControlElementTextDirection(*textFormControl);
+
+    if (auto* slotElement = dynamicDowncast<HTMLSlotElement>(element))
+        return computeSlotElementTextDirection(*slotElement);
+
+    return computeElementContainedTextAutoDirection(element);
+}
+
+static std::optional<TextDirection> computeElementTextDirection(const Element& element, TextDirectionality textDirectionality)
+{
+    // Specs: https://html.spec.whatwg.org/multipage/dom.html#attr-dir
+    switch (textDirectionality) {
+    case TextDirectionality::LTR:
+        return TextDirection::LTR;
+    case TextDirectionality::RTL:
+        return TextDirection::RTL;
+
+    case TextDirectionality::Auto:
+        // Specs: Return the auto directionality of element.
+        // Specs: If directionality is null, then return 'ltr'.
+        return computeElementAutoTextDirection(element).value_or(TextDirection::LTR);
+
+    case TextDirectionality::Undefined:
+        // Specs: If element is a bdi element, then return the auto directionality of element.
+        // Specs: If directionality is null, then return 'ltr'.
+        if (is<HTMLBDIElement>(element))
+            return computeElementAutoTextDirection(element).value_or(TextDirection::LTR);
+
+        // Specs: If element is an input element whose type attribute is in the Telephone state
+        // then return LTR;
+        if (auto* input = dynamicDowncast<HTMLInputElement>(element); input && input->isTelephoneField())
+            return TextDirection::LTR;
+
+        // Specs: If the parent is a shadow root, then return the directionality of parent's host.
+        // Specs: If the parent is an element, then return the directionality of parent.
+        if (RefPtr parent = element.parentOrShadowHostElement())
+            return computeElementTextDirection(*parent);
+
+        return std::nullopt;
+    }
+
+    return std::nullopt;
+}
+
+std::optional<TextDirection> computeElementTextDirection(const Element& element)
+{
+    return computeElementTextDirection(element, elementTextDirectionality(element));
+}
+
+static void updateElementSelfTextDirection(Element& element, std::optional<TextDirection> direction, Element* exclude = nullptr)
+{
+    bool usesTextDirection = !!direction;
+    auto effectiveDirection = direction.value_or(TextDirection::LTR);
+
+    Style::PseudoClassChangeInvalidation styleInvalidation(element, CSSSelector::PseudoClass::Dir, Style::PseudoClassChangeInvalidation::AnyValue);
+    element.setUsesTextDirection(usesTextDirection);
+    element.setTextDirection(effectiveDirection);
+
+    if (RefPtr shadowRoot = element.shadowRoot()) {
+        for (Ref child : childrenOfType<Element>(*shadowRoot)) {
+            if (child.ptr() == exclude)
```

Comment:
This `exclude` business is very confusing. Why is this needed?

Reply:
`updateElementTextDirection()` for element _element_ may call `updateElementAncestorTextDirection()` which calls `updateElementTextDirection()` back for the first ancestor _parent_ which has an auto directionality. `updateElementTextDirection()` for element _parent_ will call `updateElementDescendantsTextDirection()`. One of its descents might be _element_. So we may end up updating the textDirection bits for the element we started with _element_. `exclude` will help skipping the initiating element will going up in the tree and then going down.